Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,349 @@
+        }
+            
+        // Events
+
+        public sealed override Task OnUserConnected(IUser user)
+        {
+            return OnThingAdded(GlobalUserListeners, SingleUserListeners, UserListeners, user, user.SessionId, user.UserId);
+        }
+
+        public sealed override Task OnChannelCreated(IChannel channel)
+        {
+            return OnThingAdded(GlobalChannelListeners, SingleChannelListeners, ChannelListeners, channel, channel.ChannelId, channel.ChannelId);
+        }
+
+        private async Task OnThingAdded<T>(ConcurrentBag<GlobalListener<T>> globalListeners,
+                ConcurrentDictionary<int, IList<GlobalListener<T>>> singleListeners,
+                ConcurrentDictionary<int, IList<ActiveListener<T>>> listeners,
+                T thing, int uniqueId, int permanentId) where T : IReadModifyWriteObject
+        {
+            // whenever a thing (e.g. user) is added, all registered listeners for this thing must be added
+            // first get all registered users by getting all global listeners and registered single listeners
+            foreach (var gl in globalListeners.Concat(singleListeners[permanentId])) {
+                // create the tuple to add (mapper, listener, mapper(thing))
+                var al = new ActiveListener<T>(gl);
+                // add each of them to the listeners list
+                // if it is the first one, a new List must be created and the object be added
+                // otherwise add it to the list
+                listeners.AddOrUpdate(uniqueId,
+                    // add new list
+                    i =>
+                    {
+                        var l = new List<ActiveListener<T>>();
+                        l.Add(al);
+                        return l;
+                    },
+                    // update list
+                    (i, l) =>
+                    {
+                        l.Add(al);
+                        return l;
+                    }
+                );
+            }
+
+            // execute all listeners async
+            await Task.WhenAll(listeners[uniqueId].Select(t => t.Listener(thing)));
+            // save changes
+            await thing.SaveChanges();
+        }
+
+        public sealed override Task OnUserDisconnected(IUser user)
+        {
+            return OnThingRemoved(UserListeners, user.SessionId);
+        }
+
+        public override Task OnChannelDeleted(IChannel channel)
+        {
+            return OnThingRemoved(ChannelListeners, channel.ChannelId);
+        }
+
+        private Task OnThingRemoved<T>(ConcurrentDictionary<int, IList<ActiveListener<T>>> listeners, int uniqueId) where T : IReadModifyWriteObject
+        {
+            IList<ActiveListener<T>> dontCare;
+            listeners.TryRemove(uniqueId, out dontCare);
```

Comment:
If this can't fail, we should throw if it does, just for sanity.

If it *can* fail but we genuinely don't care, a comment would be nice.